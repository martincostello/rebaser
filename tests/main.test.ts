// Copyright (c) Martin Costello, 2023. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import * as core from '@actions/core';
import { afterAll, beforeAll, describe, expect, test } from '@jest/globals';
import { ActionFixture } from './ActionFixture';

describe('rebaser', () => {
  const rebaseTimeout = 15000;
  const runFixture = async (name: string = ''): Promise<ActionFixture> => {
    const fixture = new ActionFixture();
    await fixture.initialize();

    if (name) {
      await fixture.setupRepositoryFromFixture(name);
    }

    await fixture.run();

    return fixture;
  };

  describe('when global.json has conflicts', () => {
    let fixture: ActionFixture;

    beforeAll(async () => {
      fixture = await runFixture('global.json');
    }, rebaseTimeout);

    afterAll(async () => {
      await fixture?.destroy();
    });

    test('generates no errors', () => {
      expect(core.error).toHaveBeenCalledTimes(0);
      expect(core.setFailed).toHaveBeenCalledTimes(0);
    });

    test('outputs the correct result', () => {
      expect(fixture.getOutput('result')).toBe('success');
    });

    test('rebases the branch', async () => {
      expect(await fixture.commitHistory(3)).toEqual(['Apply target', 'Apply patch', 'Apply base']);
    });

    test('matches the snapshot', async () => {
      expect(await fixture.getFileContent('global.json')).toMatchSnapshot();
    });
  });

  describe('when Directory.Packages.props has conflicts', () => {
    let fixture: ActionFixture;

    beforeAll(async () => {
      fixture = await runFixture('Directory.Packages.props');
    }, rebaseTimeout);

    afterAll(async () => {
      await fixture?.destroy();
    });

    test('generates no errors', () => {
      expect(core.error).toHaveBeenCalledTimes(0);
      expect(core.setFailed).toHaveBeenCalledTimes(0);
    });

    test('outputs the correct result', () => {
      expect(fixture.getOutput('result')).toBe('success');
    });

    test('rebases the branch', async () => {
      expect(await fixture.commitHistory(3)).toEqual(['Apply target', 'Apply patch', 'Apply base']);
    });

    test('matches the snapshot', async () => {
      expect(await fixture.getFileContent('Directory.Packages.props')).toMatchSnapshot();
    });
  });

  describe('when package.json has conflicts', () => {
    let fixture: ActionFixture;

    beforeAll(async () => {
      fixture = await runFixture('package.json');
    }, rebaseTimeout * 2);

    afterAll(async () => {
      await fixture?.destroy();
    });

    test('generates no errors', () => {
      expect(core.error).toHaveBeenCalledTimes(0);
      expect(core.setFailed).toHaveBeenCalledTimes(0);
    });

    test('outputs the correct result', () => {
      expect(fixture.getOutput('result')).toBe('success');
    });

    test('rebases the branch', async () => {
      expect(await fixture.commitHistory(3)).toEqual(['Apply target', 'Apply patch', 'Apply base']);
    });

    test('regenerates the lock file', async () => {
      expect(await fixture.diff(3)).toContain('package-lock.json');
    });

    test('matches the snapshot', async () => {
      expect(await fixture.getFileContent('package.json')).toMatchSnapshot();
    });
  });

  describe('when C# project file has conflicts', () => {
    let fixture: ActionFixture;

    beforeAll(async () => {
      fixture = await runFixture('Project.csproj');
    }, rebaseTimeout);

    afterAll(async () => {
      await fixture?.destroy();
    });

    test('generates no errors', () => {
      expect(core.error).toHaveBeenCalledTimes(0);
      expect(core.setFailed).toHaveBeenCalledTimes(0);
    });

    test('outputs the correct result', () => {
      expect(fixture.getOutput('result')).toBe('success');
    });

    test('rebases the branch', async () => {
      expect(await fixture.commitHistory(3)).toEqual(['Apply target', 'Apply patch', 'Apply base']);
    });

    test('matches the snapshot', async () => {
      expect(await fixture.getFileContent('Project/Project.csproj')).toMatchInlineSnapshot(`
"<Project>
  <ItemGroup>
    <PackageVersion Include="System.Text.Json" Version="8.0.0-preview.6.23329.7" />
  </ItemGroup>
</Project>
"
`);
    });
  });

  describe('when a solution has multiple conflicts', () => {
    let fixture: ActionFixture;

    beforeAll(async () => {
      fixture = await runFixture('Complex');
    }, rebaseTimeout);

    afterAll(async () => {
      await fixture?.destroy();
    });

    test('generates no errors', () => {
      expect(core.error).toHaveBeenCalledTimes(0);
      expect(core.setFailed).toHaveBeenCalledTimes(0);
    });

    test('outputs the correct result', () => {
      expect(fixture.getOutput('result')).toBe('success');
    });

    test('rebases the branch', async () => {
      expect(await fixture.commitHistory(3)).toEqual(['Apply target', 'Apply patch', 'Apply base']);
    });

    test('SDK matches the snapshot', async () => {
      expect(await fixture.getFileContent('global.json')).toMatchInlineSnapshot(`
"{
  "sdk": {
    "version": "8.0.100-preview.6.23330.14"
  }
}
"
`);
    });

    test('packages matches the snapshot', async () => {
      expect(await fixture.getFileContent('Directory.Packages.props')).toMatchSnapshot();
    });
  });

  describe('when branch is up-to-date', () => {
    let fixture: ActionFixture;

    beforeAll(async () => {
      fixture = await runFixture('global.json');
      fixture.reset();

      await fixture.run();
    }, rebaseTimeout);

    afterAll(async () => {
      await fixture?.destroy();
    });

    test('generates no errors', () => {
      expect(core.error).toHaveBeenCalledTimes(0);
      expect(core.setFailed).toHaveBeenCalledTimes(0);
    });

    test('outputs the correct result', () => {
      expect(fixture.getOutput('result')).toBe('upToDate');
    });

    test('the branch is still rebased from before', async () => {
      expect(await fixture.commitHistory(3)).toEqual(['Apply target', 'Apply patch', 'Apply base']);
    });

    test('matches the snapshot', async () => {
      expect(await fixture.getFileContent('global.json')).toMatchSnapshot();
    });
  });

  describe('when the conflicts cannot be resolved', () => {
    let fixture: ActionFixture;

    beforeAll(async () => {
      fixture = await runFixture('Unresolvable');
    }, rebaseTimeout);

    afterAll(async () => {
      await fixture?.destroy();
    });

    test('generates no errors', () => {
      expect(core.error).toHaveBeenCalledTimes(0);
      expect(core.setFailed).toHaveBeenCalledTimes(0);
    });

    test('outputs the correct result', () => {
      expect(fixture.getOutput('result')).toBe('conflicts');
    });

    test('aborts the rebase', async () => {
      expect(await fixture.commitHistory(3)).toEqual(['Apply target', 'Apply base', 'Initial commit']);
    });

    test('matches the snapshot', async () => {
      expect(await fixture.getFileContent('Directory.Packages.props')).toMatchSnapshot();
    });
  });
});
