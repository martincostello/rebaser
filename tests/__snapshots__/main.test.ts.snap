// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rebaser when C# project file has conflicts matches the snapshot 1`] = `
"<Project>
  <ItemGroup>
    <PackageVersion Include="System.Text.Json" Version="8.0.0-preview.6.23329.7" />
  </ItemGroup>
</Project>
"
`;

exports[`rebaser when Directory.Packages.props has conflicts from an MSBuild property matches the snapshot 1`] = `
"<Project>
  <PropertyGroup>
    <DotNetVersion>8.0.0-preview.6.23329.7</DotNetVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageVersion Include="System.Text.Json" Version="$(DotNetVersion)" />
  </ItemGroup>
</Project>
"
`;

exports[`rebaser when Directory.Packages.props has conflicts matches the snapshot 1`] = `
"<Project>
  <ItemGroup>
    <PackageVersion Include="System.Text.Json" Version="8.0.0-preview.6.23329.7" />
  </ItemGroup>
</Project>
"
`;

exports[`rebaser when a line is added packages matches the snapshot 1`] = `
"<Project>
  <ItemGroup>
    <PackageVersion Include="coverlet.msbuild" Version="6.0.0" />
    <PackageVersion Include="GitHubActionsTestLogger" Version="2.3.2" />
    <PackageVersion Include="MartinCostello.Logging.XUnit" Version="0.3.0" />
    <PackageVersion Include="Microsoft.AspNetCore.Mvc.Testing" Version="8.0.0-preview.6.23329.11" />
    <PackageVersion Include="Microsoft.EntityFrameworkCore.Sqlite" Version="8.0.0-preview.6.23329.4" />
    <PackageVersion Include="Microsoft.Extensions.TimeProvider.Testing" Version="8.0.0-preview.6.23360.2" />
    <PackageVersion Include="Microsoft.NET.Test.Sdk" Version="17.7.0" />
    <PackageVersion Include="Microsoft.Playwright" Version="1.36.0" />
    <PackageVersion Include="Microsoft.TypeScript.MSBuild" Version="5.1.5" />
    <PackageVersion Include="MockingLibrary" Version="4.18.4" />
    <PackageVersion Include="ReportGenerator" Version="5.1.23" />
    <PackageVersion Include="Shouldly" Version="4.2.1" />
    <PackageVersion Include="StyleCop.Analyzers" Version="1.2.0-beta.507" />
    <PackageVersion Include="xunit" Version="2.5.0" />
    <PackageVersion Include="xunit.runner.visualstudio" Version="2.5.0" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="StyleCop.Analyzers" PrivateAssets="All" />
  </ItemGroup>
</Project>
"
`;

exports[`rebaser when a line is deleted packages matches the snapshot 1`] = `
"<Project>
  <ItemGroup>
    <PackageVersion Include="System.Text.Json" Version="8.0.0-preview.6.23329.7" />
  </ItemGroup>
</Project>
"
`;

exports[`rebaser when a prerelease has conflicts matches the snapshot 1`] = `
"{
  "sdk": {
    "version": "9.0.100-rc.1.24413.1"
  }
}
"
`;

exports[`rebaser when a prerelease has conflicts matches the snapshot 2`] = `
"<Project Sdk="Microsoft.NET.Sdk.Web">
  <PropertyGroup>
    <EnableConfigurationBindingGenerator>true</EnableConfigurationBindingGenerator>
    <GenerateDocumentationFile>true</GenerateDocumentationFile>
    <IsPackable>false</IsPackable>
    <NoWarn>$(NoWarn);CA1019;CA1050;CA1813;CS1591</NoWarn>
    <OpenApiDocumentsDirectory>$([System.IO.Path]::Combine($(ArtifactsPath), 'openapi'))</OpenApiDocumentsDirectory>
    <OpenApiGenerateDocuments>true</OpenApiGenerateDocuments>
    <OpenApiGenerateDocumentsOnBuild>true</OpenApiGenerateDocumentsOnBuild>
    <OpenApiGenerateDocumentsOptions>--openapi-version OpenApi3_0</OpenApiGenerateDocumentsOptions>
    <RootNamespace>TodoApp</RootNamespace>
    <TargetFramework>net9.0</TargetFramework>
    <TypeScriptCompileBlocked>true</TypeScriptCompileBlocked>
    <TypeScriptToolsVersion>latest</TypeScriptToolsVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.OpenApi" Version="9.0.0-rc.1.24413.4" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="9.0.0-rc.1.24402.2" />
    <PackageReference Include="Microsoft.Extensions.ApiDescription.Server" Version="9.0.0-rc.1.24413.4" PrivateAssets="all" />
    <PackageReference Include="Microsoft.OpenApi" Version="1.6.19" />
    <PackageReference Include="Microsoft.TypeScript.MSBuild" Version="5.5.3" PrivateAssets="all" />
    <PackageReference Include="NSwag.AspNetCore" Version="14.1.0" />
    <PackageReference Include="Swashbuckle.AspNetCore" Version="6.7.3" />
    <PackageReference Include="Swashbuckle.AspNetCore.Annotations" Version="6.7.3" />
  </ItemGroup>
  <ItemGroup>
    <Content Update="package.json;package-lock.json;tsconfig.json" CopyToPublishDirectory="Never" />
    <None Remove="scripts\\ts\\**\\*.ts" />
    <TypeScriptCompile Include="scripts\\ts\\**\\*.ts" />
  </ItemGroup>
  <Target Name="BundleAssets" BeforeTargets="BeforeBuild">
    <Exec Command="npm ci" Condition=" !Exists('$(MSBuildThisFileDirectory)\\node_modules') AND '$(GITHUB_ACTIONS)' != '' " />
    <Exec Command="npm install" Condition=" !Exists('$(MSBuildThisFileDirectory)\\node_modules') AND '$(GITHUB_ACTIONS)' == '' " />
    <Exec Command="npm run build" Condition=" !Exists('$(MSBuildThisFileDirectory)\\wwwroot\\static\\js\\main.js') " />
  </Target>
</Project>
"
`;

exports[`rebaser when a solution has multiple conflicts SDK matches the snapshot 1`] = `
"{
  "sdk": {
    "version": "8.0.100-preview.6.23330.14"
  }
}
"
`;

exports[`rebaser when a solution has multiple conflicts packages matches the snapshot 1`] = `
"<Project>
  <ItemGroup>
    <!-- A comment added in the patch -->
    <PackageVersion Include="Alexa.NET" Version="1.22.0" />
    <PackageVersion Include="Amazon.Lambda.Core" Version="2.1.0" />
    <PackageVersion Include="Amazon.Lambda.Logging.AspNetCore" Version="3.1.0" />
    <PackageVersion Include="Amazon.Lambda.RuntimeSupport" Version="1.8.8" />
    <PackageVersion Include="Amazon.Lambda.Serialization.Json" Version="2.1.1" />
    <PackageVersion Include="Amazon.Lambda.TestUtilities" Version="2.0.0" />
    <PackageVersion Include="AWSSDK.Lambda" Version="3.7.200.1" />
    <PackageVersion Include="coverlet.msbuild" Version="6.0.0" />
    <PackageVersion Include="GitHubActionsTestLogger" Version="2.3.2" />
    <PackageVersion Include="JustEat.HttpClientInterception" Version="4.0.0" />
    <PackageVersion Include="MartinCostello.Logging.XUnit" Version="0.3.0" />
    <PackageVersion Include="MartinCostello.Testing.AwsLambdaTestServer" Version="0.7.1" />
    <PackageVersion Include="Microsoft.ApplicationInsights" Version="2.21.0" />
    <PackageVersion Include="Microsoft.Extensions.DependencyInjection" Version="8.0.0-preview.6.23329.7" />
    <!-- A comment in the middle of the conflict range -->
    <PackageVersion Include="Microsoft.Extensions.Http" Version="8.0.0-preview.6.23329.7" />
    <PackageVersion Include="Microsoft.Extensions.Http.Polly" Version="8.0.0-preview.6.23329.11" />
    <PackageVersion Include="Microsoft.Extensions.Logging" Version="8.0.0-preview.6.23329.7" />
    <PackageVersion Include="Microsoft.ICU.ICU4C.Runtime" Version="72.1.0.1" />
    <PackageVersion Include="Microsoft.NET.Test.Sdk" Version="17.6.3" />
    <PackageVersion Include="MockingLibrary" Version="4.18.4" />
    <PackageVersion Include="Polly" Version="7.2.4" />
    <PackageVersion Include="Refit" Version="7.0.0" />
    <PackageVersion Include="ReportGenerator" Version="5.1.23" />
    <PackageVersion Include="Shouldly" Version="4.2.1" />
    <PackageVersion Include="StyleCop.Analyzers" Version="1.2.0-beta.507" />
    <PackageVersion Include="System.Text.Json" Version="8.0.0-preview.6.23329.7" />
    <PackageVersion Include="xunit" Version="2.5.0" />
    <PackageVersion Include="xunit.runner.visualstudio" Version="2.5.0" />
    <PackageVersion Include="Xunit.SkippableFact" Version="1.4.13" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="StyleCop.Analyzers" PrivateAssets="All" />
  </ItemGroup>
  <ItemGroup Condition=" '$(IsTestProject)' == 'true' ">
    <PackageReference Include="coverlet.msbuild" PrivateAssets="All" />
    <PackageReference Include="ReportGenerator" PrivateAssets="All" />
  </ItemGroup>
</Project>
"
`;

exports[`rebaser when a the number of lines in the diff is uneven packages matches the snapshot 1`] = `
"<Project>
  <ItemGroup>
    <PackageVersion Include="coverlet.msbuild" Version="6.0.0" />
    <PackageVersion Include="GitHubActionsTestLogger" Version="2.3.2" />
    <PackageVersion Include="MartinCostello.Logging.XUnit" Version="0.3.0" />
    <PackageVersion Include="Microsoft.Extensions.DependencyInjection" Version="8.0.0-preview.6.23329.7" />
    <PackageVersion Include="Microsoft.Extensions.Http" Version="8.0.0-preview.6.23329.7" />
    <PackageVersion Include="Microsoft.Extensions.Http.Polly" Version="8.0.0-preview.6.23329.11" />
    <PackageVersion Include="Microsoft.NET.Test.Sdk" Version="17.7.0" />
    <PackageVersion Include="MockingLibrary" Version="4.18.4" />
    <PackageVersion Include="Newtonsoft.Json" Version="13.0.1" Condition=" '$(IsTestProject)' != 'true' " />
    <PackageVersion Include="Newtonsoft.Json" Version="13.0.3" Condition=" '$(IsTestProject)' == 'true' " />
    <PackageVersion Include="Polly" Version="7.2.4" />
    <PackageVersion Include="ReportGenerator" Version="5.1.23" />
    <PackageVersion Include="Shouldly" Version="4.2.1" />
    <PackageVersion Include="StyleCop.Analyzers" Version="1.1.118" />
    <PackageVersion Include="xunit" Version="2.5.0" />
    <PackageVersion Include="xunit.runner.visualstudio" Version="2.5.0" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="StyleCop.Analyzers" PrivateAssets="All" />
  </ItemGroup>
  <ItemGroup Condition=" '$(IsTestProject)' == 'true' ">
    <PackageReference Include="coverlet.msbuild" PrivateAssets="All" />
    <PackageReference Include="ReportGenerator" PrivateAssets="All" />
  </ItemGroup>
</Project>
"
`;

exports[`rebaser when branch is up-to-date matches the snapshot 1`] = `
"{
  "sdk": {
    "version": "8.0.100-preview.6.23330.14"
  }
}
"
`;

exports[`rebaser when global.json has conflicts matches the snapshot 1`] = `
"{
  "sdk": {
    "version": "8.0.100-preview.6.23330.14"
  }
}
"
`;

exports[`rebaser when package.json has conflicts matches the snapshot 1`] = `
"{
  "name": "rebaser",
  "version": "1.0.0",
  "private": true,
  "dependencies": {
    "@microsoft/signalr": "^8.0.0-preview.6.23329.11"
  }
}
"
`;

exports[`rebaser when the conflicts cannot be resolved matches the snapshot 1`] = `
"<Project>
  <ItemGroup>
    <PackageVersion Include="Microsoft.Extensions.TimeProvider.Testing" Version="8.0.0-preview.6.23360.2" />
    <PackageVersion Include="System.Text.Json" Version="8.0.0-preview.6.23329.7" />
  </ItemGroup>
</Project>
"
`;
